lib/puppet/face/facts.rb

```diff
@@ -35,4 +35,48 @@
 
   deactivate_action(:destroy)
   deactivate_action(:search)
+
+  action(:upload) do
+    summary _("Upload local facts to the puppet master.")
+    description <<-'EOT'
+      Reads facts from the local system using the `facter` terminus, then
+      saves the returned facts using the rest terminus.
+    EOT
+    returns "Nothing."
+    notes <<-'EOT'
+      This action requires that the puppet master's `auth.conf` file
+      allow save access to the `facts` REST terminus. Puppet agent does
+      not use this facility, and it is turned off by default. See
+      <https://puppet.com/docs/puppetserver/latest/config_file_auth.html>
+      for more details.
+    EOT
+    examples <<-'EOT'
+      Upload facts:
+
+      $ puppet facts upload
+    EOT
+
+    render_as :json
+
+    when_invoked do |options|
+      # Use `agent` sections  settings for certificates, Puppet Server URL,
+      # etc. instead of `user` section settings.
+      Puppet.settings.preferred_run_mode = :agent
+      Puppet::Node::Facts.indirection.terminus_class = :facter
+
+      facts = Puppet::Node::Facts.indirection.find(Puppet[:node_name_value])
+      unless Puppet[:node_name_fact].empty?
+        Puppet[:node_name_value] = facts.values[Puppet[:node_name_fact]]
+        facts.name = Puppet[:node_name_value]
+      end
+
+      Puppet::Node::Facts.indirection.terminus_class = :rest
+      server = Puppet::Node::Facts::Rest.server
+      Puppet.notice(_("Uploading facts for '%{node}' to: '%{server}'") % {
+                    node: Puppet[:node_name_value],
+                    server: server})
+
+      Puppet::Node::Facts.indirection.save(facts)
```

Comment:
@Sharpie can the caller try to pass trusted facts here? Does puppetserver and/or puppetdb terminus reject the request, or does it strip them out?

Reply:
That is a very good question. I'll look into what happens on the server side.

Reply:
Similar question, is the timestamp in the uploaded facts ignored? Or does the puppetdb terminus always ignore it and set the "server-specified" timestamp to avoid clock skew issues?

Reply:
@Sharpie have you had a chance to look into Josh's questions?

Reply:
Not yet. I've been tied up for the past two days with a customer perf problem and MCollective issues :(

Reply:
Here is the list of facts passed to the save call on line 79, we don't appear to have any of the trusted facts here:

```
#<Puppet::Node::Facts:0x007fc78e116a40 @name="localhost",
 @values={
 "puppetversion"=>"5.3.5",
 "kernelversion"=>"17.4.0",
 "uptime"=>"7 days",
 "kernelrelease"=>"17.4.0",
 "uptime_seconds"=>681530,
 "path"=>"/Users/eric.delaney/work/my_repos/puppetserver/ruby/puppet/.bundle/ruby/2.2.0/bin:/Users/eric.delaney/.rbenv/versions/2.2.8/bin:/usr/local/Cellar/rbenv/1.1.1/libexec:./ruby/puppet/bin:./ruby/facter/bin:/Users/eric.delaney/.rbenv/shims:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/puppetlabs/bin:/usr/local/munki:/Users/eric.delaney/bin:/Users/eric.delaney/.rvm/bin",
 "kernel"=>"Darwin",
 "zzz_test"=>"data",
 "facter"=>".puppetlabs",
 "xxx_test"=>"thing",
 "test"=>"testing",
 "hostname"=>"MacBook-Pro-2",
 "productname"=>"MacBookPro11,4",
 "operatingsystem"=>"Darwin",
 "processors"=>{"count"=>8,
 "speed"=>"2.2 GHz"},
 "uptime_days"=>7,
 "kernelmajversion"=>"17.4",
 "timezone"=>"PST",
 "ps"=>"ps auxwww",
 "virtual"=>"physical",
 "is_virtual"=>false,
 "architecture"=>"x86_64",
 "hardwaremodel"=>"x86_64",
 "os"=>{"name"=>"Darwin",
 "family"=>"Darwin",
 "release"=>{"major"=>"17",
 "minor"=>"4",
 "full"=>"17.4.0"}},
 "network_lo0"=>"127.0.0.0",
 "network_en0"=>"10.0.24.0",
 "rubysitedir"=>"/Users/eric.delaney/.rbenv/versions/2.2.8/lib/ruby/site_ruby/2.2.0",
 "sp_smc_version_system"=>"2.29f24",
 "sp_boot_rom_version"=>"MBP114.0177.B00",
 "sp_cpu_type"=>"Intel Core i7",
 "sp_current_processor_speed"=>"2.2 GHz",
 "sp_l2_cache_core"=>"256 KB",
 "sp_l3_cache"=>"6 MB",
 "sp_machine_model"=>"MacBookPro11,4",
 "sp_machine_name"=>"MacBook Pro",
 "sp_number_processors"=>4,
 "sp_packages"=>1,
 "sp_physical_memory"=>"16 GB",
 "sp_platform_uuid"=>"EC54FA6D-EB0B-5DE3-B6B0-81D73DF53409",
 "sp_serial_number"=>"C02T50DPG8WL",
 "sp_boot_mode"=>"normal_boot",
 "sp_boot_volume"=>"Macintosh HD",
 "sp_kernel_version"=>"Darwin 17.4.0",
 "sp_local_host_name"=>"MacBook Pro (2)",
 "sp_os_version"=>"macOS 10.13.3 (17D102)",
 "sp_secure_vm"=>"secure_vm_enabled",
 "sp_system_integrity"=>"integrity_enabled",
 "sp_uptime"=>"up 7:21:19:20",
 "sp_user_name"=>"Eric Delaney (eric.delaney)",
 "macosx_productname"=>"Mac OS X",
 "macosx_productversion"=>"10.13.3",
 "macosx_buildversion"=>"17D102",
 "macosx_productversion_major"=>"10.13",
 "macosx_productversion_minor"=>"3",
 "operatingsystemmajrelease"=>"17",
 "gid"=>"staff",
 "hardwareisa"=>"i386",
 "rubyversion"=>"2.2.8",
 "processorcount"=>8,
 "system_uptime"=>{"seconds"=>681530,
 "hours"=>189,
 "days"=>7,
 "uptime"=>"7 days"},
 "macaddress"=>"18:65:90:cb:91:23",
 "osfamily"=>"Darwin",
 "id"=>"eric.delaney",
 "uptime_hours"=>189,
 "facterversion"=>"2.5.1",
 "memorysize"=>"16.00 GB",
 "memoryfree"=>"82.46 MB",
 "swapsize"=>"2.00 GB",
 "swapfree"=>"1.15 GB",
 "swapsize_mb"=>"2048.00",
 "swapfree_mb"=>"1181.00",
 "memorysize_mb"=>"16384.00",
 "memoryfree_mb"=>"82.46",
 "swapencrypted"=>true,
 "netmask"=>"255.255.252.0",
 "domain"=>"local",
 "interfaces"=>"lo0,gif0,stf0,XHC20,en0,p2p0,awdl0,en1,en2,bridge0,utun0",
 "ipaddress_lo0"=>"127.0.0.1",
 "netmask_lo0"=>"255.0.0.0",
 "mtu_lo0"=>16384, "mtu_gif0"=>1280, "mtu_stf0"=>1280, "mtu_xhc20"=>0, "ipaddress_en0"=>"10.0.26.32",
 "macaddress_en0"=>"18:65:90:cb:91:23",
 "netmask_en0"=>"255.255.252.0",
 "mtu_en0"=>1500, "macaddress_p2p0"=>"0a:65:90:cb:91:23",
 "mtu_p2p0"=>2304, "macaddress_awdl0"=>"46:e9:09:a3:de:62",
 "mtu_awdl0"=>1484, "macaddress_en1"=>"6a:00:02:fa:21:a0",
 "mtu_en1"=>1500, "macaddress_en2"=>"6a:00:02:fa:21:a1",
 "mtu_en2"=>1500, "macaddress_bridge0"=>"6a:00:02:fa:21:a0",
 "mtu_bridge0"=>1500, "mtu_utun0"=>2000, "fqdn"=>"MacBook-Pro-2.local",
 "operatingsystemrelease"=>"17.4.0",
 "ipaddress"=>"10.0.26.32",
 "rubyplatform"=>"x86_64-darwin17",
 "clientcert"=>"localhost",
 "clientversion"=>"5.3.5",
 "clientnoop"=>false
},
@timestamp=2018-02-28 12:03:53 -0800,
@expiration=2018-02-28 12:33:53 -0800>
```

Reply:
Yeah, the `save` call wouldn't receive trusted facts unless there was a custom fact that produced a value called `trusted`.

What should happen:

  - The `/puppet/v3/facts` API should ignore any `trusted` key in the `values` section of the PUT body.

  - If the client certificate used to make connection is valid and contains trusted data, this data should be merged in as an allowed `trusted` value.

I'd bet the current API endpoint fails both of those as it pre-dated trusted data.

Reply:
Ok, looks like the PuppetDB Facts terminus does exactly the right thing during a `save` action by consulting `Puppet.lookup(:trusted_information)` which the HTTP handler sets to info recovered from the client cert, if present:

https://github.com/puppetlabs/puppetdb/blob/5.2.0/puppet/lib/puppet/indirector/facts/puppetdb.rb#L28

However, the YAML terminus will just save whatever is submitted. This can cause some problems during compilation as certain Puppet Server actions will load data from the YAML cache rather than hitting PuppetDB.